prefect_azure/container_instance.py

```diff
@@ -0,0 +1,694 @@
+from anyio.abc import TaskStatus
+from azure.core.credentials import TokenCredential
+from azure.core.polling import LROPoller
+from azure.identity import ClientSecretCredential
+from azure.mgmt.containerinstance import ContainerInstanceManagementClient
+from azure.mgmt.containerinstance.models import (
+    Container,
+    ContainerGroup,
+    ContainerGroupRestartPolicy,
+    ContainerGroupSubnetId,
+    EnvironmentVariable,
+    GpuResource,
+    ImageRegistryCredential,
+    OperatingSystemTypes,
+    ResourceRequests,
+    ResourceRequirements,
+)
+from azure.mgmt.resource import ResourceManagementClient
+from prefect.docker import get_prefect_image_name
+from prefect.infrastructure.base import Infrastructure, InfrastructureResult
+from prefect.utilities.asyncutils import run_sync_in_worker_thread, sync_compatible
+from pydantic import Field, SecretStr
+from typing_extensions import Literal
+
+from prefect_azure.credentials import ContainerInstanceCredentials
+
+ACI_DEFAULT_CPU = 1.0
+ACI_DEFAULT_MEMORY = 1.0
+ACI_DEFAULT_GPU = 0.0
+DEFAULT_CONTAINER_ENTRYPOINT = "/opt/prefect/entrypoint.sh"
+
+
+class ContainerGroupProvisioningState(str, Enum):
+    """
+    Terminal provisioning states for ACI container groups. Per the Azure docs,
+    the states in this Enum are the only ones that can be relied on as dependencies.
+    """
+
+    SUCCEEDED = "Succeeded"
+    FAILED = "Failed"
+
+
+class ContainerRunState(str, Enum):
+    """
+    Terminal run states for ACI containers.
+    """
+
+    RUNNING = "Running"
+    TERMINATED = "Terminated"
+
+
+class ContainerInstanceJobResult(InfrastructureResult):
+    """
+    The result of an `ContainerInstanceJob` run.
+    """
+
+
+class ContainerInstanceJob(Infrastructure):
+    """
+    <span class="badge-api experimental"/>
+    Run a command using a container on Azure Container Instances.
+    Note this block is experimental. The interface may change without notice.
+    """
+
+    _block_type_slug = "container-instance-job"
+    _block_type_name = "Azure Container Instance Job"
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/6AiQ6HRIft8TspZH7AfyZg/39fd82bdbb186db85560f688746c8cdd/azure.png?h=250"  # noqa
+    _description = "Run tasks using Azure Container Instances. Note this block is experimental. The interface may change without notice."  # noqa
+
+    type: Literal["container-instance-job"] = Field(
+        default="container-instance-job", description="The slug for this task type."
+    )
+    aci_credentials: ContainerInstanceCredentials
+    resource_group_name: str = Field(
+        default=...,
+        title="Azure Resource Group Name",
+        description=(
+            "The name of the Azure Resource Group in which to run Prefect ACI tasks."
+        ),
+    )
+    subscription_id: SecretStr = Field(
+        default=...,
+        title="Azure Subscription ID",
+        description="The ID of the Azure subscription to create containers under.",
+    )
```

Comment:
Other Azure credentials blocks have `subscription_id` as a field. This seems auth related, so it might make sense to move this field to the container instance credentials block.

Reply:
I originally had it in the credentials block, but moved it into the infra block because Azure uses subscription IDs for billing and resource tagging/tracking...and authentication, but it's not a 1:1 relationship between user and subscription.   A service/security principal can spin up containers and other resources across multiple subscriptions they have access to. And they often do, at least in my experience working with Azure at other companies. 

I've seen multiple subscriptions used to split things up by team, by department, by project, by type of workload, by dev/staging/production, or even into separate subscriptions for different stages of a data pipeline. [Microsoft's guidance on this](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/decision-guides/subscriptions/) isn't exactly clear. 

In many of these scenarios, the same user/service identities deploy resources like containers across multiple subscriptions. Azure's APIs (including container instances) reinforce this approach by keeping credentials and subscriptions separate. You don't use subscription ID to authenticate; you only need it to provide when you make API calls to create resources (and then Azure checks with Active Directory to make sure you're authorized to make the API request). So, I ended up moving subscription to where it is now, because that matches the expectations of developers accustomed to working with Azure. 

Having said that, I'm 100% happy to move subscription ID to the credential block if that's the best fit for Prefect. I just wanted to share the 10,000 foot overview of how I've seen orgs and developers interact with Azure subscriptions.

Reply:
Thanks for that excellent perspective! It seems like including the subscription ID with the credentials block would make the credentials blocks less reusable. I think your choice of including the subscription ID on the container instance block makes a lot of sense considering the common uses that you described. 

Since the subscription ID is also used in authentication, I think the pattern that would work would be to allow a subscription ID to be passed into `get_client`methods on credentials blocks. This would allow different subscription IDs to be used with the same credentials block. Would that approach work with the ways that you've seen Azure subscriptions used in the past?